src/misc/unicode.cpp

```diff
@@ -32,416 +32,618 @@
+		// Not a combining mark or empty grapheme
+		Invalidate();
+		return;
+	}
+
+	if (std::find(marks.begin(), marks.end(), code_point) != marks.end()) {
+		// Combining mark already present
+		return;
+	}
+
+	marks.push_back(code_point);
+	marks_sorted.push_back(code_point);
+	std::sort(marks_sorted.begin(), marks_sorted.end());
+}
 
-    if (mark_1st && mark_2nd)
-        return; // up to 2 combining marks supported, ignore the rest
+void Grapheme::StripMarks()
+{
+	marks.clear();
+	marks_sorted.clear();
+}
 
-    if (mark_1st)
-        mark_2nd = code_point;
-    else
-        mark_1st = code_point;
+void Grapheme::Decompose()
+{
+	if (!is_valid || is_empty) {
+		// Can't decompose invalid or empty grapheme
+		return;
+	}
+
+	while (decomposition_rules.count(code_point) != 0) {
+		const auto& rule = decomposition_rules.at(code_point);
+		code_point       = rule.code_point;
+		for (const auto mark : rule.marks) {
+			AddMark(mark);
+		}
+	}
 }
 
-void Grapheme::StripMarks()
+bool Grapheme::operator==(const Grapheme& other) const
 {
-    mark_1st = 0;
-    mark_2nd = 0;
+	return (is_empty == other.is_empty) && (is_valid == other.is_valid) &&
+	       (code_point == other.code_point) &&
+	       (marks_sorted == other.marks_sorted);
 }
 
-bool Grapheme::operator<(const Grapheme &other) const
+bool Grapheme::operator<(const Grapheme& other) const
 {
-    // TODO: after migrating to C++20 try to use a templated lambda here
-
-    if (code_point < other.code_point)
-        return true;
-    else if (code_point > other.code_point)
-        return false;
-
-    if (mark_1st < other.mark_1st)
-        return true;
-    else if (mark_1st > other.mark_1st)
-        return false;
-
-    if (mark_2nd < other.mark_2nd)
-        return true;
-    else if (mark_2nd > other.mark_2nd)
-        return false;
-
-    if (!is_empty && other.is_empty)
-        return true;
-    else if (is_empty && !other.is_empty)
-        return false;
-
-    if (!is_valid && other.is_valid)
-        return true;
-    else if (is_valid && !other.is_valid)
-        return false;
-
-    // '*this == other' condition is, unfortunately, not recognized by older compilers
-    assert(code_point == other.code_point &&
-           mark_1st   == other.mark_1st &&
-           mark_2nd   == other.mark_2nd &&
-           is_empty   == other.is_empty &&
-           is_valid   == other.is_valid);
-
-    return false;
+	if (code_point < other.code_point) {
+		return true;
+	} else if (code_point > other.code_point) {
+		return false;
+	}
+
+	if (marks_sorted.empty() && other.marks_sorted.empty()) {
+		return false;
+	}
+
+	if (marks_sorted.size() < other.marks_sorted.size()) {
+		return true;
+	} else if (marks_sorted.size() > other.marks_sorted.size()) {
+		return false;
+	}
+
+	for (size_t idx = 0; idx < marks_sorted.size(); ++idx) {
+		if (marks_sorted[idx] < other.marks_sorted[idx]) {
+			return true;
+		} else if (marks_sorted[idx] > other.marks_sorted[idx]) {
+			return false;
+		}
```

Comment:
The `else if` can become a stand-alone `if` because the prior `if` returns.

Reply:
Done.